Initial implementation

romalytvynenko opened the pull request:
Initial implementation of UI with event manager for flexible rendering.

romalytvynenko committed:
core that register UI and templates rendering via event manager

romalytvynenko committed:
templates and composer config

romalytvynenko committed:
add registering pages, screen options and menu entries

romalytvynenko committed:
fix namespace and caps

romalytvynenko committed:
fix module type

mecha commented on src/TemplateManager.php:
```diff
@@ -0,0 +1,166 @@
+     */
+    public function __construct(ContainerInterface $c, $prefix = 'eddbk')
+    {
+        $this->eventManager = $c->get('event-manager');
```
Since the container is not saved in this class, consider accepting an `EventManagerInterface` as the first parameter rather than a `ContainerInterface`. The `TemplateManager` can then be instantiated as:

```
new TemplateManager($c->get('event_manager'), $prefix);
```

mecha commented on src/TemplateManager.php:
```diff
@@ -0,0 +1,166 @@
+    protected function _register($templates)
+    {
+        foreach ($templates as $template) {
+            $templateFilePath = __DIR__ . '/../templates/' . $template . '.phtml';
```
I'd rather avoid using relative paths that traverse upwards in the file system. I recommend adding a constant in the `module.php` file that records the module's directory, such as

```
define('WP_BOOKINGS_UI_MODULE_DIR', __DIR__);
```

Then use this constant here.

mecha commented on src/UIModule.php:
```diff
@@ -0,0 +1,87 @@
+    public function setup()
+    {
+        return $this->_createContainer([
+            'template-manager' => function (ContainerInterface $c) {
```
As per our recent agreement, here and wherever else necessary, change the service keys declared to `snake_case`. Also assume that services provided by other modules are also in `snake_case`.

mecha commented on src/UIModule.php:
```diff
@@ -0,0 +1,87 @@
+    public function run(ContainerInterface $c = null)
+    {
+        /* @var $eventManager EventManagerInterface */
+        $eventManager = $c->get('event-manager');
```
As per our recent agreement, here and wherever else necessary, change the service keys declared to `snake_case`. Also assume that services provided by other modules are also in `snake_case`.

mecha commented on module.php:
```diff
@@ -0,0 +1,8 @@
+use RebelCode\Bookings\WordPress\Module\UIModule;
+
+return function (ContainerInterface $c) {
+    return new UIModule($c->get('container-factory'));
```
As per our recent agreement, here and wherever else necessary, change the service keys declared to `snake_case`. Also assume that services provided by other modules are also in `snake_case`.

mecha commented on src/UIModule.php:
```diff
@@ -0,0 +1,202 @@
+            $containerFactory,
+            'bookings-ui-module',
+            ['wp-events'],
+            $this->_loadPhpConfigFile(__DIR__ . '/../config.php')
```
I'd rather avoid using relative paths that traverse upwards in the file system. I recommend adding a constant in the `module.php` file that records the module's directory, such as

```
define('WP_BOOKINGS_UI_MODULE_DIR', __DIR__);
```

Then use this constant here.

mecha commented on src/TemplateManager.php:
```diff
@@ -0,0 +1,166 @@
+    protected function _getGeneralTemplates()
+    {
+        return [
+            'components',
```
I think the template lists are better kept in the configuration file(s). These methods can be kept and made to return the lists from the config.

mecha commented on src/UIModule.php:
```diff
@@ -0,0 +1,202 @@
+     * @throws \Psr\Container\ContainerExceptionInterface
+     * @throws \Psr\Container\NotFoundExceptionInterface
+     */
+    protected function _renderMetabox($c)
```
I think it makes more sense for this method, as well as the other similar methods (`_renderMainPage()`, `_renderSettingsPage()`, `_renderAboutPage`) to accept a `TemplateManager` instance instead of the container. The less the container is passed around, the better. I would only make exceptions if a class/method would otherwise need 5 or more parameters.

mecha commented on src/UIModule.php:
```diff
@@ -0,0 +1,202 @@
+use Psr\EventManager\EventManagerInterface;
+use RebelCode\Modular\Module\AbstractBaseModule;
+
+class UIModule extends AbstractBaseModule
```
As much as possible, try to keep the module class name similar to the module package and repo name. Also, stick to camel case. In this case, I would call this class `WpBookingsUiModule`.

romalytvynenko committed:
requested pr changes

romalytvynenko commented:
@mecha, fixed everything according  your comments

mecha committed:
Fixed module dependency name

mecha committed:
Fixed event manager service key

mecha committed:
Added line ending at end of file

mecha committed:
Merge branch 'develop' into task/initial-implementation

mecha commented:
@romalytvynenko Be advised, the package has been renamed and the repository has been moved.

mecha committed:
Fixed README badges

romalytvynenko committed:
working UI templates assembling

romalytvynenko committed:
initial assets shipping

romalytvynenko committed:
working config

romalytvynenko committed:
build instructions

romalytvynenko committed:
add correct hook for assets

romalytvynenko committed:
new event management

romalytvynenko committed:
fix assets path

mecha commented on src/WpBookingsUiModule.php:
```diff
@@ -0,0 +1,201 @@
+        $this->_initModule(
+            $containerFactory,
+            'wp_bookings_ui',
+            ['wp_event_manager'],
```
This module is not going to be used anymore.

mecha commented on src/WpBookingsUiModule.php:
```diff
@@ -0,0 +1,201 @@
+    public function run(ContainerInterface $c = null)
+    {
+        /* @var EventManagerInterface $eventManager  */
+        $eventManager = $c->get('wp_event_manager');
```
Here and everywhere else, depend on `event_manager` instead.

mecha commented on src/WpBookingsUiModule.php:
```diff
@@ -0,0 +1,201 @@
+            function () use ($templateManager) {
+                return $this->_renderMetabox($templateManager);
+            },
+            $this->_getConfig()['metabox']['post_type']
```
Kindly optimize this code. `$this->_getConfig()` is being called 3 times and each time it is being accessed at `'metabox'`. Consider retrieving `$this->_getConfig()['metabox']` only once.

mecha commented on src/WpBookingsUiModule.php:
```diff
@@ -0,0 +1,201 @@
+            $this->_getConfig()['menu']['about']['page_title'],
+            $this->_getConfig()['menu']['about']['menu_title'],
+            $this->_getConfig()['menu']['about']['capability'],
+            $this->_getConfig()['menu']['about']['menu_slug'],
```
Kindly optimize this code. `$this->_getConfig()` is being called 16 times, 8 of which it is being accessed at `'root'`, 4 at `'settings'` and 4 at `'about'`.

Consider retrieving `$this->_getConfig()` only once and accessing `'root'`, `'settings'` and `'about'` as required.

romalytvynenko committed:
using EventsConsumerTrait

romalytvynenko committed:
remove wp_event_manager module dependency, depend on event_manager across calls`

romalytvynenko committed:
add detailed explanation how front is runned and which stuff is required to be available on backend

romalytvynenko committed:
readme fix

romalytvynenko committed:
readme fix json

romalytvynenko committed:
fixed conflicts

romalytvynenko committed:
Merge pull request #2 from RebelCode/task/assets-shipping

Assets shipping

romalytvynenko committed:
service metabox integration with data and ui

romalytvynenko committed:
working bookings fetching using endpoints

romalytvynenko committed:
bookings data fetched from endpoints

romalytvynenko committed:
fixed major UI problems related to working with async data

romalytvynenko committed:
fix missing loaders

romalytvynenko committed:
UX issues

romalytvynenko committed:
fix issue

romalytvynenko committed:
enqueue assets from container

XedinUnknown commented on config.php:
```diff
@@ -0,0 +1,90 @@
+    'assets' => [
+        'require' => 'https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.5/require.js',
+
+        'ui-module::assets/js/main.js' => plugins_url(WP_BOOKINGS_UI_MODULE_RELATIVE_DIR . '/assets/js/main.js', EDDBK_FILE),
```
Like I mentioned before, the paths need to be relative to an abstract plugins' root. `::` is not a valid directory separator, and therefore this is wrong.

XedinUnknown commented on src/WpBookingsUiModule.php:
```diff
@@ -0,0 +1,344 @@
+        $templateManager = $c->get('template_manager');
+
+        $assetsContainer = $this->_getContainerFactory()->make([
+            'definitions' => $this->_getConfig()['assets']
```
This looks wrong, because it assumes that the config is an array. Entries should be accessed by key. If this requires extra implementation, use `_containerGet()` or `_containerGetPath()`.

XedinUnknown commented on config.php:
```diff
@@ -0,0 +1,90 @@
+            'position' => 20
+        ],
+        'settings' => [
+            'page_title' => __('Settings', 'eddbk'),
```
Data in the config must be static. This assumes the existence of the `__()` function, and invokes it, which should never happen. On the other hand, such strings definitely need to be i18n-ized. @mecha, what do you think we can do about this?

romalytvynenko committed:
fixing required changes

romalytvynenko committed:
enable bookings tick is working now

romalytvynenko committed:
cursor pointer on event hover

romalytvynenko committed:
creating bookings using calendar

romalytvynenko committed:
add all time units to session length

romalytvynenko committed:
add bookings filter option to search by clients on both bookings views

romalytvynenko committed:
added statuses filter to both bookings views

romalytvynenko committed:
update expected endpoint documentation

XedinUnknown commented on src/WpBookingsUiModule.php:
```diff
@@ -0,0 +1,356 @@
+
+    use NormalizeKeyCapableTrait;
+
+    private $bookingsPageId;
```
Why is this `private`, @romalytvynenko?

XedinUnknown commented on src/WpBookingsUiModule.php:
```diff
@@ -0,0 +1,356 @@
+     * @param $eventFactory
+     * @throws \Dhii\Exception\InternalException
+     */
+    public function __construct($key, ContainerFactoryInterface $containerFactory, $eventManager, $eventFactory)
```
Why no types declared for `$eventManager` and `$eventFactory`?

XedinUnknown commented on src/WpBookingsUiModule.php:
```diff
@@ -0,0 +1,356 @@
+    {
+        return $this->_createContainer([
+            'template_manager' => function () {
+                $templateManager = new TemplateManager($this->eventManager, $this->eventFactory);
```
~Why is a class being instantiated here?~

Because this is a service definition.

romalytvynenko committed:
fix sessions holder so it hold only 10 items and the adds scroll

romalytvynenko committed:
fix width of price input

romalytvynenko committed:
add message for empty sessions state

romalytvynenko committed:
fix availability datepickers to disable certain dates

romalytvynenko committed:
fixed availabilities issues

romalytvynenko committed:
bookings status style fi

romalytvynenko committed:
fix button switcher style

romalytvynenko committed:
select width fix

romalytvynenko committed:
style fixes for buttons and editor fixes

romalytvynenko committed:
notification about overlapping availabilities

romalytvynenko committed:
form validation

romalytvynenko committed:
fluid close confirmation

romalytvynenko committed:
simple copy on ui build

romalytvynenko committed:
update gitignore

romalytvynenko committed:
fix endpoints config

romalytvynenko committed:
new modular system and comments

romalytvynenko committed:
working modular config

romalytvynenko committed:
style and UI fixes

romalytvynenko committed:
fixed all UI issues

romalytvynenko committed:
update booking js

romalytvynenko committed:
fix timepicker

romalytvynenko committed:
provide sprintf formatter

romalytvynenko committed:
fixing mobiles layout problems

romalytvynenko committed:
fied all mobile issues

romalytvynenko committed:
better api integration. understand get, can work with clients, can process new statuses

romalytvynenko committed:
statuses now are saving in the options

romalytvynenko committed:
booking options serialized

romalytvynenko committed:
moved data retrieving from dummy module

romalytvynenko committed:
all fixes for modals and tablet layout

romalytvynenko committed:
all fixes from lists

romalytvynenko committed:
updates according last config changes

romalytvynenko committed:
remove unneeded array normalization

romalytvynenko committed:
removed unneeded use of array normalization

XedinUnknown commented on src/WpBookingsUiModule.php:
```diff
@@ -71,12 +71,18 @@ public function __construct(
     {
         return $this->_setupContainer(
-            $this->_loadPhpConfigFile(WP_BOOKINGS_UI_MODULE_DIR . '/config.php'),
+            $this->_loadPhpConfigFile(WP_BOOKINGS_UI_MODULE_CONFIG_FILE),
```
This is one of the places where you need to map paths, not URLs.

XedinUnknown commented on src/WpBookingsUiModule.php:
```diff
@@ -71,12 +71,18 @@ public function __construct(
                     return $templateManager;
+                },
+                'assets_urls_map' => function () {
+                    $assetsUrlsMap = require_once WP_BOOKINGS_UI_MODULE_CONFIG_DIR . '/assets_urls_map.php';
```
Another case where you need to map a path.

XedinUnknown commented on config/config.php:
```diff
@@ -0,0 +1,93 @@
+    'metabox' => [
+        'id' => 'service_booking_settings',
+        'post_type' => 'post', //'download',
+        'title' => __('Booking Options', 'eddbk'),
```
I understand that right now we do not have a perfect solution for i18n in configs. But at least, use the constant for text domain. Alternatively, as discussed before, the consuming code can do the i18n instead: otherwise there is no way to interpolate placeholder values.

romalytvynenko replied:
Fixed

XedinUnknown commented on config/config.php:
```diff
@@ -0,0 +1,93 @@
+<?php
+
+return [
```
In general, the format of this config appears wrong. Configuration should all follow a uniform structure, relative to the absolute config root. In your case, a consumer in any module or hook can do `$c->get('metabox')`, and will get the metabox configuration from line 8. But this is wrong: what is this metabox for? Why is it in the global config?

Instead, the configuration should put module specific config under a module namespace, and leave only generic configuration out. Ideally, however, there should not be any "generic" configuration that exists on top of any module configuration. Please update accordingly.

romalytvynenko replied:
Fixed

XedinUnknown replied:
Looks almost great. When can you fix this to only do i18n in the consuming code?

romalytvynenko committed:
fixed config with text domain and module domain

XedinUnknown commented on config/config.php:
```diff
@@ -0,0 +1,95 @@
+        ],
+        'assets' => [
+            'require.js' => 'require',
+            'bookings' => [
```
The key looks like it needs to be changed. What does it mean for those paths to be in `bookings`?

XedinUnknown commented on config/config.php:
```diff
@@ -0,0 +1,95 @@
+            'require.js' => 'require',
+            'bookings' => [
+                'app.min.js' => 'bookings_ui/dist/app.min.js',
+                'main.js' => 'bookings_ui/assets/js/main.js',
```
~Why is this a map? Does the key matter? Where and how is it used?~

XedinUnknown replied:
Apparently, the key matters, because it is used as a handle for enqueuing assets with WP. Handles are also used to declare one asset's dependency on another.

romalytvynenko committed:
moved config i18n to consumer

XedinUnknown commented on package.json:
```diff
@@ -0,0 +1,20 @@
+    "postinstall": "cp ./node_modules/booking-js/dist/js/app.min.js ./dist/app.min.js"
+  },
+  "devDependencies": {
+    "booking-js": "git@github.com:RebelCode/bookings-js.git#11bc8c09ea0d6f52ed5e587297c8ddb3bf456fcc",
```
If possible, declare dependency on a tag instead of commit. Even better - use a constraint according to SemVer.

romalytvynenko committed:
build on separate command

romalytvynenko committed:
update .gitignore

romalytvynenko committed:
vagrant added

romalytvynenko committed:
update gitignore

romalytvynenko committed:
updated vagrant to install php5.6, curl, xml

romalytvynenko committed:
new availability logic

romalytvynenko committed:
moved booking options on dowmload page

XedinUnknown committed:
Now using Trusty environment

This is necessary because target PHP version is 5.4.
ALso, the VM now has same name as package.

XedinUnknown committed:
Changed VM IP address

This will help to prevent clashes with other VMs

XedinUnknown committed:
Changed provisioning algorithm

- Updating and upgrading first.
- Now using PHPBrew to install PHP.
- Now installing PHP 5.4 instead of 5.6: this is the target version.
- Installing Git.
- Other configuration for Git and SSH and PHP build.

XedinUnknown committed:
The `booking-js` JS dep is no longer explicitly requested from a URL

XedinUnknown committed:
Added Phing build file

This will attempt to build the project. Right now, this means:

1. Create a `dist` directory if it doesn't already exist.
2. Clean that directory, in case there are files.
3. If npm packages are not installed, install them.
4. Build `booking-js` Node module directly into the `dist` directory.

XedinUnknown committed:
Build task now depends on clean and prepare

XedinUnknown committed:
Fixing npm build comand

romalytvynenko committed:
fixed symlink problems

XedinUnknown committed:
Removed extra code from Vagrantfile

XedinUnknown committed:
Phing will now build JS modules from a list

XedinUnknown committed:
Added ability to create symlinks inside VM

Requires Vagrant to be run as administrator

XedinUnknown committed:
Added dependency caching to VM

XedinUnknown committed:
Merge branch 'task/initial-implementation' into task/build-procedure-and-environment

XedinUnknown commented on src/WpBookingsUiModule.php:
```diff
@@ -0,0 +1,500 @@
+use Dhii\Event\EventFactoryInterface;
+use Psr\Container\ContainerInterface;
+use Psr\EventManager\EventManagerInterface;
+use RebelCode\EddBookings\RestApi\Controller\ControllerInterface;
```
This interface comes from [rebelcode/rcmod-eddbk-rest-api][], but that package is not a dependency of this module.

[rebelcode/rcmod-eddbk-rest-api]: https://packagist.org/packages/rebelcode/rcmod-eddbk-rest-api

XedinUnknown committed:
No longer depending on global WP function `__()`

XedinUnknown committed:
Merge branch 'task/initial-implementation' into task/build-procedure-and-environment

XedinUnknown commented on src/WpBookingsUiModule.php:
```diff
@@ -0,0 +1,500 @@
+     *
+     * @var
+     */
+    public $bookingsPageId;
```
Why are there public properties?

XedinUnknown replied:
And the type is undocumented.

XedinUnknown committed:
Removed old commented out debugging code

XedinUnknown commented on src/WpBookingsUiModule.php:
```diff
@@ -0,0 +1,500 @@
+
+        $assetsConfig = $c->get('assets_urls_map');
+
+        $eventManager->attach('admin_enqueue_scripts', function () use ($assetsConfig, $c) {
```
Why is there a hard dependency on an event manager? Use `AbstractBaseModule#_attach()`.

XedinUnknown committed:
Decresed coupling to event manager in certain places

romalytvynenko committed:
availability now can be longer than one day

romalytvynenko committed:
Merge branch 'task/initial-implementation' of https://github.com/RebelCode/rcmod-wp-bookings-ui into task/initial-implementation

romalytvynenko committed:
session tab is first now

romalytvynenko committed:
bump booking-js version

romalytvynenko committed:
new input name

romalytvynenko committed:
bump booking js version

romalytvynenko committed:
update caps

romalytvynenko committed:
client can be `null`

XedinUnknown committed:
Now installing Composer in VM

Also, redundant steps will be skipped

XedinUnknown committed:
Merged `task/initial-implementation`

XedinUnknown committed:
Merge branch 'task/initial-implementation' of https://github.com/RebelCode/rcmod-wp-bookings-ui into task/build-procedure-and-environment

Conflicts:
	package.json

XedinUnknown committed:
Merge pull request #3 from RebelCode/task/build-procedure-and-environment

Build procedure and environment

XedinUnknown commented on src/WpBookingsUiModule.php:
```diff
@@ -0,0 +1,499 @@
+use RebelCode\Modular\Module\AbstractBaseModule;
+use Dhii\Util\String\StringableInterface as Stringable;
+
+class WpBookingsUiModule extends AbstractBaseModule
```
Missing docs.

XedinUnknown commented on src/WpBookingsUiModule.php:
```diff
@@ -0,0 +1,499 @@
+
+class WpBookingsUiModule extends AbstractBaseModule
+{
+    use ContainerGetCapableTrait;
```
Missing `@since` docs.

XedinUnknown commented on src/WpBookingsUiModule.php:
```diff
@@ -0,0 +1,499 @@
+     *
+     * @param string|Stringable $key The module key.
+     * @param string[]|Stringable[] $dependencies The module  dependencies.
+     * @param ContainerFactoryInterface $configFactory The config factory.
```
Looks like CS fixer was not run on this project.

XedinUnknown commented on src/WpBookingsUiModule.php:
```diff
@@ -0,0 +1,499 @@
+    }
+
+    /**
+     * @inheritdoc
```
Missing `@since` tag.

XedinUnknown replied:
In many places.

XedinUnknown commented on src/WpBookingsUiModule.php:
```diff
@@ -0,0 +1,499 @@
+     *
+     * @return bool
+     */
+    protected function _onAppPage()
```
`_isOnAppPage()` please, etc. `_onAppPage()` looks like an event handler.

XedinUnknown commented on src/WpBookingsUiModule.php:
```diff
@@ -0,0 +1,499 @@
+     */
+    protected function _getServiceAppState()
+    {
+        $pageId = get_post()->ID;
```
Try to abstract everything that uses WP.

XedinUnknown commented on src/WpBookingsUiModule.php:
```diff
@@ -0,0 +1,499 @@
+     * @param EventManagerInterface $eventManager
+     * @param TemplateManager $templateManager
+     */
+    protected function _adminInit($eventManager, $templateManager, $c)
```
Why does this method need event manager instance? Check other methods too.

XedinUnknown commented on src/WpBookingsUiModule.php:
```diff
@@ -0,0 +1,499 @@
+        $settingsMenuConfig = $c->get('wp_bookings_ui/menu/settings');
+        $aboutMenuConfig = $c->get('wp_bookings_ui/menu/about');
+
+        $this->bookingsPageId = add_menu_page(
```
Can we make sure that the booking page ID is never requested before it is set?

XedinUnknown replied:
@romalytvynenko, why do you need to ask WP for the page ID? Can you use something constant instead, like  `$menuSlug` (the 4th param)?

romalytvynenko replied:
I'll check...

romalytvynenko replied:
WP adds some stuff in front of slug:
`toplevel_page_eddbk-bookings` when slug is `eddbk-bookings`

XedinUnknown replied:
I understand. But why do you need it? Can you not rely on it, and rely on something that *you* are in control of?

romalytvynenko replied:
Got it, I'll check

romalytvynenko replied:
I can check it using:
`strpos(get_current_screen()->id, $menuSlug) !== false`

Or: 
`$_GET['page'] === $menuSlug`, but I know :(

XedinUnknown commented on src/WpBookingsUiModule.php:
```diff
@@ -0,0 +1,499 @@
+            $this->__($rootMenuConfig->get('menu_title')),
+            $rootMenuConfig->get('capability'),
+            $rootMenuConfig->get('menu_slug'),
+            function () use ($templateManager) {
```
Abstract all usage of template manager. The class should be able to just render a template in consumer code.

romalytvynenko committed:
added missing docs

XedinUnknown commented on src/WpBookingsUiModule.php:
```diff
@@ -0,0 +1,499 @@
+            /*
+             * All available statuses in application.
+             */
+            'statuses' => $this->_trigger('eddbk_bookings_statuses', [
```
This stuff looks like it can go into a config

mecha replied:
I actually have a module that provides the statuses as config.

https://github.com/RebelCode/rcmod-eddbk-booking-logic/blob/task/initial-module-code/config.php#L9

However, they're not mapped to translatable human friendly strings.

XedinUnknown replied:
@romalytvynenko, make sure that all logic uses a single source of truth, which is obviously the booking module.

XedinUnknown commented on src/WpBookingsUiModule.php:
```diff
@@ -0,0 +1,499 @@
+     * @param $endpointsConfig
+     * @return array
+     */
+    protected function _prepareEndpoints($endpointsConfig)
```
This looks like quite generic code that can, based on a config structure convention, work for any module. @mecha, thoughts?

mecha replied:
Not sure what exactly this code is trying to achieve. Why does the UI module need to register WP REST API endpoints?

romalytvynenko replied:
@mecha, this method is not for registering endpoints. Using this method I'm preparing endpoints for using in JS. I cannot wrap it in the `rest_url(..)` inside `config.php` so I'm doing it in the module.

XedinUnknown commented on src/WpBookingsUiModule.php:
```diff
@@ -0,0 +1,499 @@
+        $screenOptions = get_user_option($c->get('wp_bookings_ui/screen_options/key'), $user->ID);
+        if (!$screenOptions) {
+            return [
+                "in_cart", "draft", "pending", "approved", "rejected", "scheduled", "cancelled", "completed"
```
Something else that can be retrieved from config instead?

XedinUnknown commented on src/WpBookingsUiModule.php:
```diff
@@ -0,0 +1,499 @@
+            $rootMenuConfig->get('position')
+        );
+
+        add_submenu_page(
```
All this stuff can be achieved by using [rebelcode/wp-admin-menu][] and [rebelcode/wp-admin-page]. It will also allow you to move the definitions of the menus and pages into service definitions.

[rebelcode/wp-admin-menu]: https://github.com/RebelCode/wp-admin-menu
[rebelcode/wp-admin-page]: https://github.com/RebelCode/wp-admin-page

XedinUnknown commented on src/WpBookingsUiModule.php:
```diff
@@ -0,0 +1,499 @@
+     *
+     * @param TemplateManager $templateManager
+     * @return mixed
+     * @throws \Psr\Container\ContainerExceptionInterface
```
Why no import?

XedinUnknown commented on src/WpBookingsUiModule.php:
```diff
@@ -0,0 +1,499 @@
+        return $this->_setupContainer(
+            $this->_loadPhpConfigFile(WP_BOOKINGS_UI_MODULE_CONFIG_FILE),
+            [
+                'template_manager' => function ($c) {
```
Is it possible to move service definitions into a different place? Maybe just a separate PHP file?

mecha replied:
Like a secondary config file?

XedinUnknown replied:
Yeaz, totally.

XedinUnknown replied:
Maybe, no need to use Config implementation for it. Just simply move.

romalytvynenko committed:
new services app state event

romalytvynenko committed:
statuses translations from config

romalytvynenko committed:
moved knowledge about configs key away from meethod

romalytvynenko commented:
@XedinUnknown done (Accept statuses and the map of known labels) here: https://github.com/RebelCode/rcmod-wp-bookings-ui/commit/bb843a0c486df4f3e27fd696be3e936bd1d27d70

romalytvynenko committed:
moved away controller and now relying on event for getting list of services

XedinUnknown commented on src/WpBookingsUiModule.php:
```diff
@@ -228,6 +221,20 @@ protected function _getTranslatedStatuses($statuses, $statusesLabels)
+     *
+     * @since [*next-version*]
+     *
+     * @return mixed List of all services.
```
Why is the type so wide? Can it return an integer?

romalytvynenko replied:
Got it. I think that `Traversable` is good, can I use it instead?

XedinUnknown replied:
I think you should ask @mecha what he will return. I guess that it will be quite wide. But of course not that wide :D

mecha replied:
To be honest, its just `array`, because the transformer that I'm using in the event normalizes to array @XedinUnknown So still a pretty vague type.

XedinUnknown replied:
I would say, allow `array`, as well as `Traversable`, and `stdClass`. But an array of what? @mecha.

mecha replied:
Prior to transforming, an `array|stdClass|Traversable` of arrays. There is no formal model object for services (yet). After transforming, an array of arrays.

XedinUnknown commented on src/WpBookingsUiModule.php:
```diff
@@ -170,16 +169,12 @@ protected function _isOnBookingsPage()
              */
-            'statuses' => $this->_trigger('eddbk_bookings_statuses', [
-                'statuses' => $this->_getTranslatedStatuses($c)
+            'statuses' => $this->_trigger('eddbk_bookings_translated_statuses', [
```
This looks weird. The app state will be like this?

```javascript
{
	statuses: {
		statuses: [
			
		]
	}
}
```

romalytvynenko replied:
No man, as you remember, this is the event (you can see in the line's end `...->getParam('statuses')`. So state will be like this:
```
{
  'statuses': {
    ...
    'draft': 'Draft',
    ...
  }
}

XedinUnknown replied:
Yep, true, my bad.

XedinUnknown commented on src/WpBookingsUiModule.php:
```diff
@@ -0,0 +1,558 @@
+     *
+     * @since [*next-version*]
+     *
+     * @return array|mixed|object
```
Return description missing.

XedinUnknown replied:
An array containing what?

XedinUnknown replied:
Can return any object?

XedinUnknown replied:
Can return integer or string?

romalytvynenko committed:
updated docs for module

romalytvynenko committed:
statuses moved avay from methods

romalytvynenko committed:
moved services definitions into separate file

romalytvynenko committed:
updates according comments

XedinUnknown commented on src/WpBookingsUiModule.php:
```diff
@@ -0,0 +1,564 @@
+     *
+     * @return array List of all services.
+     */
+    protected function _getListOfServices()
```
Just `_getServices()` is enough.

XedinUnknown commented on src/WpBookingsUiModule.php:
```diff
@@ -0,0 +1,564 @@
+     */
+    protected function _renderSettingsPage($templateManager)
+    {
+        throw new \Exception('Implement Settings page.');
```
Alias `\Exception` to `RootException`.

XedinUnknown replied:
Exception messages must be internationalized.

romalytvynenko committed:
rename excludesDates to excludeDates

romalytvynenko committed:
bump bookings-js version

romalytvynenko committed:
bumo verion and root exception

XedinUnknown commented:
At last, the menu implementation is available [here](https://github.com/RebelCode/wp-admin-menu/releases/tag/v0.1-alpha1).

romalytvynenko committed:
add missing docs

romalytvynenko committed:
add missing dev dependencies

romalytvynenko committed:
csfix applied for module code

romalytvynenko committed:
abstraction of template manager usage

XedinUnknown commented on src/WpBookingsUiModule.php:
```diff
@@ -0,0 +1,517 @@
+            }
+            $translatedStatuses[$status] = $this->_containerHas($statusesLabels, $status)
+                ? $this->__($this->_containerGet($statusesLabels, $status))
+                : $status;
```
Don't we need to translate the label? I think that was the deal.

romalytvynenko replied:
Hmm... I think that this is exactly what I'm doing here:
```
if (labels has key status code)
  return translate status code's label
else
  return status code
```

XedinUnknown replied:
But if the label is the code, as is the case where there is no known label, then what?

romalytvynenko replied:
As we discussed, we will return status code without translation. Is it wrong and I should apply translate function to any result?

XedinUnknown replied:
Yep. Actually, the deal was that you translate the code. Please do that.

XedinUnknown commented on src/WpBookingsUiModule.php:
```diff
@@ -0,0 +1,517 @@
+        /*
+         * Add screen options on bookings management page.
+         */
+        add_filter('screen_settings', function ($settings, \WP_Screen $screen) {
```
Looks like a missing import.

XedinUnknown commented on src/WpBookingsUiModule.php:
```diff
@@ -0,0 +1,517 @@
+        /*
+         * @todo: this is not working as expected (nothing happens).
+         */
+//        $this->_attach('screen_settings', function ($event) use ($templateManager) {
```
If this is solved, please remove this commented-out code.

XedinUnknown reviewed (changes requested):
Almost there :)

romalytvynenko committed:
screen_settings now added with events approach

romalytvynenko committed:
translate label and status

XedinUnknown commented on src/WpBookingsUiModule.php:
```diff
@@ -0,0 +1,510 @@
+    public function run(ContainerInterface $c = null)
+    {
+        $this->metaboxPageId   = $c->get('wp_bookings_ui/metabox/post_type');
+        $this->templateManager = $c->get('template_manager');
```
Template manager is unused in this scope. Also, not sure why you would need to store it in a property.

romalytvynenko replied:
I've stored it in the property because I use it in the method for rendering.

XedinUnknown replied:
But do you really need to access it this way? A better way would be to abstract the retrieval of the template manager, and get it from the container.

XedinUnknown commented on src/WpBookingsUiModule.php:
```diff
@@ -430,23 +430,14 @@ function () {
          * Add screen options on bookings management page.
          */
-        add_filter('screen_settings', function ($settings, \WP_Screen $screen) {
+        $this->_attach('screen_settings', function ($event) {
```
@romalytvynenko, does this actually work?

romalytvynenko replied:
Yes, it is actually working.

XedinUnknown replied:
Amazing!

romalytvynenko commented:
As far as I know, I have access to container only in the `run()` method. So
here I'm getting TemplateManager from the container and assigning to the
property to use in future.

On Thu, May 10, 2018 at 4:55 PM, Anton Ukhanev <notifications@github.com>
wrote:

> *@XedinUnknown* commented on this pull request.
> ------------------------------
>
> In src/WpBookingsUiModule.php
> <https://github.com/RebelCode/rcmod-wp-bookings-ui/pull/1#discussion_r187336761>
> :
>
> > +    {
> +        return $this->_setupContainer(
> +            $this->_loadPhpConfigFile(WP_BOOKINGS_UI_MODULE_CONFIG_FILE),
> +            $this->_getServicesDefinitions()
> +        );
> +    }
> +
> +    /**
> +     * {@inheritdoc}
> +     *
> +     * @since [*next-version*]
> +     */
> +    public function run(ContainerInterface $c = null)
> +    {
> +        $this->metaboxPageId   = $c->get('wp_bookings_ui/metabox/post_type');
> +        $this->templateManager = $c->get('template_manager');
>
> But do you really need to access it this way? A better way would be to
> abstract the retrieval of the template manager, and get it from the
> container.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/RebelCode/rcmod-wp-bookings-ui/pull/1#discussion_r187336761>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AF3mlKk-ihu85_GdKU5Vzg6Sp67fhrOsks5txEbcgaJpZM4S-6QE>
> .
>

XedinUnknown commented:
@mecha, this is a good point. Of course, this way there's a chance that the manager-consuming method will be run before the manager is available, but this appears to be inevitable. What do you think we can do?

I think that maybe later on we could delegate this stuff to another layer, like a runner, which could be initialized with the container, and then run as usual, using accessors, and everything will be guaranteed. Thoughts?

@romalytvynenko, for now leave this as is.

romalytvynenko committed:
build only css with phing

romalytvynenko committed:
fixing front warnings

romalytvynenko committed:
fixed calendar warning

romalytvynenko committed:
dev version of vue

romalytvynenko committed:
update naming of tasks

romalytvynenko committed:
fixed main.js

XedinUnknown committed:
Merge pull request #4 from RebelCode/task/booking-js-from-cdn

Booking JS now loaded from CDN

romalytvynenko committed:
use ^0.1-alpha1 of modular

romalytvynenko committed:
add ability to pass bookings enabled from server

mecha commented:
@XedinUnknown My approach has been to create a class that implements `InvocableInterface` and the `__invoke()` method. The class can accept the dependencies that it needs in its constructor and save them in its state.

romalytvynenko committed:
precise until dat

romalytvynenko committed:
add ability to display error from server

romalytvynenko committed:
fixed bookings app to work according latest api changes

romalytvynenko committed:
added template for timezone picker

romalytvynenko committed:
used timezone-picker for editor

XedinUnknown commented:
Yea, something like that. Regardless of how it gets invoked, I think we can agree that the deps should go in as constructor args, and the object deals with its own known state.

romalytvynenko committed:
client time correctly displayed according selected timezone

romalytvynenko committed:
bump build version

mecha committed:
Added a build `verbose` option for more output

romalytvynenko committed:
user can select timzone for availability now

romalytvynenko committed:
Merge branch 'task/initial-implementation' of https://github.com/RebelCode/rcmod-wp-bookings-ui into task/initial-implementation

romalytvynenko committed:
bump JS version

romalytvynenko committed:
moved date format to config, fixed booking creation issues

romalytvynenko committed:
fixed #5 statuses now works fine on new installation

romalytvynenko committed:
added config for hidden statuses, fixed #5

romalytvynenko committed:
fix working for session length tab, fixes #6

romalytvynenko committed:
bump js version

romalytvynenko committed:
add styles for timezone

romalytvynenko committed:
add tzFree config

romalytvynenko committed:
end now is not red
